File: renderer.py
from color import Color
from scene import Scene
from shapes import *

from random import random
from utilities import print_progress_bar, write_color


def find_nearest_object(r, objects):
    """Iterates through all objects in a scene and calculates the closest object the ray has hit

    Args:
        r (Ray): The current ray being cast
        objects (List[Shapes]): A list of all current objects in scene

    Returns:
        tuple(Object, float): Return tuples pair for object that is the closest, along with t-value for ray. If t-value is None, no objects were hit
    """
    distance = None
    obj = None
    for o in objects:
        t = o.hit(r)
        if t is None:
            continue
        if distance is None or t < distance:
            distance = t
            obj = o
    return obj, distance


def color_ray(r, scene, depth):
    if depth == 0:
        return Color(0, 0, 0)

    obj_hit, t = find_nearest_object(r, scene.objects)
    # if isinstance(obj_hit, Cube):
    #     x = 10
    if t is None:
        return scene.background(r.direction)

    hit_pos = r(t)
    normal = obj_hit.normal_at(hit_pos)
    color = obj_hit.material.color

    bounce_ray = obj_hit.material.bounce(r, normal, hit_pos)
    emitted = obj_hit.material.emitted()
    if not bounce_ray:
        return emitted

    return emitted + color * color_ray(bounce_ray, scene, depth - 1)


def render(scene, output='outputs/output', quality=3):
    
    HEIGHT = [100, 240, 360, 720, 2160][quality]
    MAX_DEPTH = [5, 25, 50, 50, 100][quality]
    NUM_SAMPLES = [50, 50, 100, 10000, 15000][quality]
        
    ASPECT_RATIO = 16 / 9
    WIDTH = int(HEIGHT * ASPECT_RATIO)

    with open(f"{output}.ppm", "w") as f:
        f.write(f"P3\n{WIDTH} {HEIGHT}\n255\n")

        for j in range(HEIGHT - 1, -1, -1):
            print_progress_bar((HEIGHT - j) / HEIGHT)
            for i in range(WIDTH):
                c = Color()
                for _ in range(NUM_SAMPLES):
                    u = (i + random()) / (WIDTH - 1)
                    v = (j + random()) / (HEIGHT - 1)
                    r = scene.camera.get_ray(u, v)
                    c += color_ray(r, scene, depth=MAX_DEPTH)
                c /= NUM_SAMPLES
                write_color(f, c)

File: renderer.cpp
#include <iostream>

#ifdef USE_OPENMP
#include <omp.h>
#endif

#include "materials/emissive.h"
#include "renderer.h"

Renderer::Renderer(int imageHeight, int samplesPerPixel, int maxDepth,
                   Vector background, float aspectRatio)
    : imageHeight(imageHeight),
      samplesPerPixel(samplesPerPixel),
      maxDepth(maxDepth),
      background(background),
      aspectRatio(aspectRatio) {
    imageWidth = (int)(aspectRatio * imageHeight);
}

Vector Renderer::rayColor(Ray &r, int depth) const {
    if (depth <= 0) {
        return Vector(0, 0, 0);
    }

    double t = INFINITY;
    Shape *hitShape = nullptr;

    for (auto shape : shapes) {
        double currentT = shape->hit(r);
        if (currentT > 0 && currentT < t) {
            t = currentT;
            hitShape = shape.get();
        }
    }

    if (hitShape == nullptr) {
        return background;
    }

    if (std::dynamic_pointer_cast<Emissive>(hitShape->material) !=
        nullptr) {  // Checking if hitShape material is Emissive
        // cast hitShape material to Emissive
        return std::dynamic_pointer_cast<Emissive>(hitShape->material)
            ->emitted();
    }

    Vector hitPos = r(t);
    Vector normal = hitShape->normalAt(hitPos);
    Vector color = hitShape->material->color;

    Ray bounceRay = hitShape->material->bounce(r, normal, hitPos);
    return color * rayColor(bounceRay, depth - 1);
}

void Renderer::addShape(std::shared_ptr<Shape> shape) {
    shapes.push_back(shape);
}

void Renderer::render(Camera &camera, std::string outputFilename) const {
    // Print the number of threads being used
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dist(-1, 1);

    Vector *buffer = new Vector[imageWidth * imageHeight];
#ifdef USE_OPENMP
    std::cout << "Number of threads: " << omp_get_max_threads() << std::endl;

#pragma omp parallel for schedule(dynamic, 1)
#endif
    for (int j = 0; j < imageHeight; j++) {
        fprintf(stderr, "\rRendering (%d spp) %5.2f%%", samplesPerPixel,
                100. * j / (imageHeight - 1));

        for (int i = 0; i < imageWidth; i++) {
            Vector color = Vector(0, 0, 0);
            for (int s = 0; s < samplesPerPixel; s++) {
                double u = (i + dist(gen)) / (imageWidth - 1);
                double v = (j + dist(gen)) / (imageHeight - 1);
                Ray camRay = camera.getRay(u, v);
                color += rayColor(camRay, maxDepth);
            }
            color =
                Vector(sqrt(color.x / samplesPerPixel),
                       sqrt(color.y / samplesPerPixel),
                       sqrt(color.z /
                            samplesPerPixel));  // gamma correction (gamma = 2)

            double ir = 255.999 * color.x;
            double ig = 255.999 * color.y;
            double ib = 255.999 * color.z;

            buffer[(imageHeight - j - 1) * imageWidth + i] +=
                Vector(ir, ig, ib);
        }
    }

    FILE *f = fopen(outputFilename.c_str(), "w");
    fprintf(f, "P3\n%d %d\n%d\n", imageWidth, imageHeight, 255);

    for (int i = 0; i < imageWidth * imageHeight; i++) {
        fprintf(f, "%d %d %d\n", (int)buffer[i].x, (int)buffer[i].y,
                (int)buffer[i].z);
    }

    fprintf(stderr, "\nDone!\n");
}